src/java/org/apache/cassandra/db/compaction/TimeOrderedKeyCompactionStrategy.java

```diff
@@ -0,0 +1,946 @@
+ * "msg1" (or any other clause). delete (t2, msg2) from SQ delete (t1, msg1) from SQ.
+ *
+ * <p>TODO: expand
+ */
+public class TimeOrderedKeyCompactionStrategy extends AbstractCompactionStrategy
+{
+    private static final Logger logger = LoggerFactory.getLogger(TimeOrderedKeyCompactionStrategy.class);
+
+    private final TimeOrderedKeyCompactionStrategyOptions options;
+    private final TimeWindowCompactionStrategyOptions twcsOptions;
+    private final Map<SSTableReader, SSTableStats> sstables = new HashMap<>();
+
+    private volatile int estimatedRemainingTasks;
+    private final String tableName;
+
+    public TimeOrderedKeyCompactionStrategy(ColumnFamilyStore cfs, Map<String, String> options)
+    {
+        super(cfs, options);
+        this.tableName = cfs.getTableName();
+        this.estimatedRemainingTasks = 0;
+        this.options = new TimeOrderedKeyCompactionStrategyOptions(options);
+        this.twcsOptions = this.options.twcsOptions;
+
+        if (!options.containsKey(AbstractCompactionStrategy.TOMBSTONE_COMPACTION_INTERVAL_OPTION)
+                && !options.containsKey(AbstractCompactionStrategy.TOMBSTONE_THRESHOLD_OPTION))
+        {
+            disableTombstoneCompactions = true;
+            logger.debug("Disabling tombstone compactions for TOKCS");
+        }
+        else
+        {
+            logger.debug("Enabling tombstone compactions for TOKCS");
+        }
+    }
+
+    @Override
+    @SuppressWarnings("resource") // transaction is closed by AbstractCompactionTask::execute
+    public AbstractCompactionTask getNextBackgroundTask(int gcBefore)
+    {
+        SSTablesForCompaction previousCandidate = null;
+        while (true)
+        {
+            SSTablesForCompaction candidate = getNextBackgroundSSTables(gcBefore);
+
+            if (candidate == SSTablesForCompaction.EMPTY)
+            {
+                return null;
+            }
+
+            // Already tried acquiring references without success. It means there is a race with
+            // the tracker but candidate SSTables were not yet replaced in the compaction strategy
+            // manager
+            if (candidate.equals(previousCandidate))
+            {
+                logger.warn(
+                        "Could not acquire references for compacting SSTables {} which is not a problem per se,"
+                                + "unless it happens frequently, in which case it must be reported. Will retry later.",
+                        candidate);
+                return null;
+            }
+
+            LifecycleTransaction modifier = cfs.getTracker().tryModify(candidate.sstables, OperationType.COMPACTION);
+            if (modifier != null)
+            {
+                logger.debug("{}: {}", tableName, candidate);
+                return buildCompactionTask(candidate, modifier, gcBefore);
+            }
+            previousCandidate = candidate;
```

Comment:
From what I can see, we get here only if we fail to get txn from tryModify.
In that case why do we rerun the logic to get sstables to compact? Why not retry with same sstables?

Reply:
Same sstables would not make sense here. 
tryModify will see if the candidate is eligible for compaction atomically. If not, it returns null. In such a case we store the candidate, and rerun the compaction to get a fresh candidate. If the new candidate is again equal to the previous candidate, then give up. CompactionManager will take care of calling the strategy again when any previously running compaction finished or a memtable flush happens.
compaction strategy can run in parallel to the compaction itself (to get the more parallel compaction tasks.), but if the compaction strategy just gets the same thing again and again, just stop it.